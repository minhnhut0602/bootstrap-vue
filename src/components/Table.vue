<template>
    <table class="table">
        <thead v-if="headers.length">
            <tr>
                <slot
                    v-for="header in headers"
                    :header="header"
                    :name="`header-${header}`"
                >
                    <th :key="header.key">
                        {{ header.label || header.key }}
                    </th>
                </slot>
            </tr>
        </thead>

        <tbody>
            <tr
                v-for="(item, index) in items"
                :key="`item-${index}`"
            >
                <slot
                    v-for="header in headers"
                    :name="header.key"
                    :item="item"
                >
                    <td :key="`item-${index}-${header}`">
                        {{ item[header.key] }}
                    </td>
                </slot>
            </tr>
        </tbody>
    </table>
</template>

<script lang="ts" setup>
import {PropType, computed, defineProps} from 'vue';

export interface Field {
    key: string;
    label?: string;
};

const props = defineProps({
    fields: {
        type: Array as PropType<Field[]>,
        required: false,
        default: undefined,
    },
    items: {
        type: Array as PropType<Record<string, any>>,
        default: () => [],
    },
});

const headers = computed<Field[]>(() => {
    if (props.fields) {
        return props.fields;
    }

    const first = props.items.find(i => i !== undefined);

    return first
        ? Object.keys(first)
            .map(f => ({key: f}))
        : [];
});
</script>
