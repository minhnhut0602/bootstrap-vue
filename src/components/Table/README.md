# Table

```vue
<Table :fields="[Field]" :items="[Object]" multiSort hover />
```

## Props

| Prop name   | Description                                       | Type                            | Values | Default     |
| ----------- | ------------------------------------------------- | ------------------------------- | ------ | ----------- |
| fields      | The fields of the table                           | Field[]                         | -      | undefined   |
| items       | The items that are shown in the table             | Record&lt;string, unknown&gt;[] | -      | () =&gt; [] |
| multiSort   | Defines if multiple items can be sorted at once   | boolean                         | -      | false       |
| hover       | Defines if the `table-hover` class should be used | boolean                         | -      | false       |
| onClick:row | Defines if the `table-click` class should be used | func                            | -      | undefined   |

## Slots

| Name                   | Description                                 | Bindings                                                                                                                                                                             |
| ---------------------- | ------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `header-${header.key}` | Displays the header of the table            | **key** `string` - The key of the header<br/>**header** `object` - The header object<br/>**name** `string` - The header name                                                         |
| top-row                | Displays as the top row of the table        |                                                                                                                                                                                      |
| no-items               | Displays if there are no items in the table |                                                                                                                                                                                      |
| row                    | Displays the row of the table               | **item** `object` - The item object<br/>**index** `int` - The index of the item                                                                                                      |
| header.key             | Displays the content of the table row       | **key** `any` - The key of the header<br/>**name** `any` - The name of the header<br/>**item** `object` - The item of the row<br/>**index** `int` - The index of the item of the row |
| bottom-row             | Displays as the bottom row of the table     |                                                                                                                                                                                      |

## Events

| Event name | Properties                                   | Description                                   |
| ---------- | -------------------------------------------- | --------------------------------------------- |
| click:row  | **item** `object` - The row that was clicked | Fired when a row is clicked                   |
| sort       | **sorted** `object` - The sort object        | Fired when the sorting of a column is changed |
