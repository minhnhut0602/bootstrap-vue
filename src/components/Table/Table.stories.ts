import OwTable from '.';

export default {
    title: 'Content/Table',
    component: OwTable,
    argTypes: {
    },
};

const template = (args) => ({
    components: {OwTable},
    setup() {
        return {args};
    },
    template: '<OwTable v-bind="args" />',
});

export const Default = template.bind({});

Default.args = {
    fields: [
        {key: 'firstName', label: 'First name'},
        {key: 'lastName', label: 'Last name'},
    ],
    items: [
        {firstName: 'Anakin', lastName: 'Skywalker'},
        {firstName: 'Jar Jar', lastName: 'Binks'},
    ],
};
