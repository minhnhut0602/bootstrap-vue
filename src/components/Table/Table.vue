<template>
    <table
        class="table"
        :class="classes"
    >
        <thead v-if="headers.length">
            <tr>
                <slot
                    v-for="header in headers"
                    :key="`header-${header.key}`"
                    :header="header"
                    :name="`header-${header.key}`"
                >
                    <th
                        :key="header.key"
                        :class="{
                            sortable: header.sortable,
                            asc: sorted[header.key] === 'asc',
                            desc: sorted[header.key] === 'desc',
                        }"
                        @click="sort(header)"
                    >
                        {{ header.label || header.key }}
                    </th>
                </slot>
            </tr>
        </thead>

        <tbody>
            <tr
                v-for="(item, index) in items"
                :key="`item-${index}`"
                @click="$emit('click:row', item)"
            >
                <slot
                    v-for="header in headers"
                    :key="header.key"
                    :name="header.key"
                    :item="item"
                >
                    <td :key="`item-${index}-${header}`">
                        {{ item[header.key] }}
                    </td>
                </slot>
            </tr>
        </tbody>
    </table>
</template>

<script lang="ts">
export interface Field {
    key: string;
    label?: string;
    sortable?: boolean;
    asc?: boolean;
    desc?: boolean;
}
</script>

<script lang="ts" setup>
import {PropType, computed, reactive} from 'vue';
import {useClasses} from '@/composables';

const props = defineProps({
    fields: {
        type: Array as PropType<Field[]>,
        required: false,
        default: undefined,
    },
    items: {
        type: Array as PropType<Record<string, unknown>[]>,
        default: () => [],
    },
    hover: {
        type: Boolean,
        default: false,
    },
    click: {
        type: Boolean,
        default: false,
    },
});

const emits = defineEmits(['click:row', 'sort']);

const sorted = reactive(
    (props.fields || [])
        .filter(f => f.sortable && (f.asc || f.desc))
        .reduce((obj, item) => Object.assign(obj, {[item.key]: item.asc ? 'asc' : 'desc'}), {}),
);

const headers = computed<Field[]>(() => {
    if (props.fields) {
        return props.fields;
    }

    const first = props.items.find(i => i !== undefined);

    return first
        ? Object.keys(first)
            .map(f => ({key: f}))
        : [];
});

const sort = (field: Field) => {
    const key = field.key;

    const sort = sorted[key];

    if (field.sortable) {
        if (sort === 'asc') {
            sorted[key] = 'desc';
        } else if (sort === 'desc') {
            delete sorted[key];
        } else {
            sorted[key] = 'asc';
        }
    }

    emits('sort', sorted);
};

const {classes} = useClasses(computed(() => [
    props.hover && 'table-hover',
    props.click && 'table-click',
]));
</script>
