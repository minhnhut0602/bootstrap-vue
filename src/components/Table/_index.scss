.table {
    th.sortable {
        cursor: pointer;

        @each $sort in ['asc', 'desc'] {
            &.#{$sort}::after {
                background-image: escape-svg($table-sort-icon);
                content: '';
                height: 1rem;
                margin: 0 0.5rem;
                position: absolute;
                width: 1rem;

                @if($sort == 'desc') {
                    transform: rotate(180deg);
                }
            }
        }
    }
}

.table-click {
    tbody > tr {
        cursor: pointer;
    }
}
