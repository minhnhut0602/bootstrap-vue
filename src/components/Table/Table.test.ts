import Table from '.';
import {shallowMount} from '@vue/test-utils';

const props = {
    items: [
        {id: 1, name: 'Word'},
        {id: 2, name: 'Press'},
        {id: 3, name: 'ISDABEST'},
        {id: 4, name: '#prank'},
    ],
};

const fields = [
    {label: 'ajdie', key: 'id', sortable: true, sort: 'asc'},
    {label: 'neem', key: 'name'},
];

describe('template', () => {
    componentRenderTest(Table);

    componentRenderTest(Table, {props}, true, 'renders with items');

    componentRenderTest(
        Table,
        {
            props: {
                ...props,
                fields,
            },
        },
        true,
        'renders with fields and items',
    );

    componentSlotRenderTest(Table, 'header-id', {props});

    componentSlotRenderTest(Table, 'row', {props});

    componentSlotRenderTest(Table, 'id', {props});

    componentSlotRenderTest(Table, 'no-items', {
        props: {
            items: [],
        },
    });

    componentWrapperClassTest(Table, {hover: true}, 'table-hover');

    it('emits event on click table row', async() => {
        const wrapper = shallowMount(Table, {
            props: {
                ...props,
                'onClick:row': () => 'test',
            },
        });

        expect(wrapper.classes('table-click')).toBe(true);

        await wrapper.find('tbody tr').trigger('click');

        expect(wrapper.emitted('click:row')[0]).toEqual([props.items[0]]);
    });
});

describe('headers', () => {
    it('is empty when no items nor fields have been given', () => {
        const wrapper = shallowMount(Table);

        expect(wrapper.vm.headers).toStrictEqual([]);
    });

    it('extracts from item when no fields given', () => {
        const wrapper = shallowMount(Table, {props});

        expect(wrapper.vm.headers).toStrictEqual([
            {key: 'id'},
            {key: 'name'},
        ]);
    });

    it('uses fields when given', () => {
        const wrapper = shallowMount(Table, {
            props: {fields},
        });

        expect(wrapper.vm.headers).toStrictEqual(fields);
    });
});

describe('sort', () => {
    it('sorts header if fields are sortable', async() => {
        const wrapper = shallowMount(Table, {
            props: {fields},
        });

        const firstTh = wrapper.find('th');

        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-asc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'asc'});

        await firstTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable', 'table-heading-sortable-desc']);

        expect(wrapper.vm.sorted).toStrictEqual({id: 'desc'});

        expect(wrapper.emitted('sort')[0]).toEqual([{id: 'desc'}]);

        await firstTh.trigger('click');

        expect(firstTh.classes()).toEqual(['table-heading-sortable']);

        expect(wrapper.vm.sorted).toStrictEqual({});

        expect(wrapper.emitted('sort')[0]).toEqual([{}]);
    });

    it('does not emit when field is not sortable', async() => {
        const wrapper = shallowMount(Table, {
            props: {fields: [{label: 'yooo', key: 'oooy'}]},
        });

        const firstTh = wrapper.find('th');

        await firstTh.trigger('click');

        expect(wrapper.emitted('sort')).toBeUndefined();
    });
});
