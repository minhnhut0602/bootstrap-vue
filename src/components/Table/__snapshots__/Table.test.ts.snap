// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<table class=\\"table\\">
  <!--v-if-->
  <tbody></tbody>
</table>"
`;

exports[`template renders with fields and items 1`] = `
"<table class=\\"table\\">
  <thead>
    <tr>
      <th class=\\"sortable asc\\">ajdie</th>
      <th class=\\"\\">neem</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>Word</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Press</td>
    </tr>
    <tr>
      <td>3</td>
      <td>ISDABEST</td>
    </tr>
    <tr>
      <td>4</td>
      <td>#prank</td>
    </tr>
  </tbody>
</table>"
`;

exports[`template renders with items 1`] = `
"<table class=\\"table\\">
  <thead>
    <tr>
      <th class=\\"\\">id</th>
      <th class=\\"\\">name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>Word</td>
    </tr>
    <tr>
      <td>2</td>
      <td>Press</td>
    </tr>
    <tr>
      <td>3</td>
      <td>ISDABEST</td>
    </tr>
    <tr>
      <td>4</td>
      <td>#prank</td>
    </tr>
  </tbody>
</table>"
`;
